rust-multiaddr/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+use std::fmt;
+use std::iter::FromIterator;
+use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6, IpAddr, Ipv4Addr, Ipv6Addr};
+use std::str::FromStr;
+
+/// Representation of a Multiaddr.
+#[derive(PartialEq, Eq, Clone, Hash)]
+pub struct Multiaddr {
+    bytes: Vec<u8>,
+}
+
+impl ToString for Multiaddr {
+    /// Convert a Multiaddr to a string
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use multiaddr::Multiaddr;
+    ///
+    /// let address = Multiaddr::new("/ip4/127.0.0.1/udt").unwrap();
+    /// assert_eq!(address.to_string(), "/ip4/127.0.0.1/udt");
+    /// ```
+    ///
+    fn to_string(&self) -> String {
+        let mut out = String::new();
+        for s in self.iter() {
+            out.push_str(&s.to_string());
+        }
+        out
+    }
+}
+
+impl fmt::Debug for Multiaddr {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.to_string().fmt(f)
+    }
+}
+
+impl Multiaddr {
+    /// Create a new multiaddr based on a string representation, like
+    /// `/ip4/127.0.0.1/udp/1234`.
+    ///
+    /// # Examples
+    ///
+    /// Simple construction
+    ///
+    /// ```
+    /// use multiaddr::Multiaddr;
+    ///
+    /// let address = Multiaddr::new("/ip4/127.0.0.1/udp/1234").unwrap();
+    /// assert_eq!(address.to_bytes(), [
+    ///     4, 127, 0, 0, 1,
+    ///     17, 4, 210
+    /// ]);
+    /// ```
+    ///
+    pub fn new(input: &str) -> Result<Multiaddr> {
+        let mut bytes = Vec::new();
+
+        let mut parts = input.split('/');
+        if !parts.next().ok_or(Error::InvalidMultiaddr)?.is_empty() {
+            return Err(Error::InvalidMultiaddr);
+        }
+
+        while let Some(part) = parts.next() {
+            let protocol: ProtocolId = part.parse()?;
+            let addr_component = match protocol.size() {
+                ProtocolArgSize::Fixed { bytes: 0 } => {
+                    protocol.parse_data("")?     // TODO: bad design
+                },
+                _ => {
+                    let data = parts.next().ok_or(Error::MissingAddress)?;
+                    protocol.parse_data(data)?
+                },
+            };
+
+            addr_component.write_bytes(&mut bytes).expect("writing to a Vec never fails");
+        }
+
+        Ok(Multiaddr { bytes: bytes })
+    }
+
+    /// Return a copy to disallow changing the bytes directly
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.bytes.to_owned()
+    }
+
+    /// Extracts a slice containing the entire underlying vector.
+    pub fn as_slice(&self) -> &[u8] {
+        &self.bytes
+    }
+
+    /// Return a list of protocols
+    ///
+    /// # Examples
+    ///
+    /// A single protocol
+    ///
+    /// ```
+    /// use multiaddr::{Multiaddr, ProtocolId};
+    ///
+    /// let address = Multiaddr::new("/ip4/127.0.0.1").unwrap();
+    /// assert_eq!(address.protocol(), vec![ProtocolId::IP4]);
+    /// ```
+    ///
+    #[inline]
+    pub fn protocol(&self) -> Vec<ProtocolId> {
```

Comment:
Can we make this `to_protocols` to make it obvious that it's doing work? Even better, make it `protocols` and return an `impl Iterator<Item = &ProtocolId>`

Reply:
Deprecated the function instead